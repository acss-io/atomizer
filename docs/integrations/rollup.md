---
description: A simple guide to setup Atomizer with Rollup.
layout: docs
section: docs
title: Rollup
---

## Install the plugin

A [Rollup](https://rollupjs.org/) plugin is available as part of the [`atomizer-plugins`](https://github.com/acss-io/atomizer/tree/main/packages/atomizer-plugins) package. Install the package in your project first.

```shell
npm i atomizer-plugins -D
```

## Create your Atomizer config

Make sure Atomizer can parse your project files in your `atomizer.config.js` file.

```js
module.exports = {
    content: [
        './src/**/*.{html,js}',
    ],
}
```

## Update the config

Then configure your `rollup.config.js` to include the plugin.

```js
import { rollup } from 'atomizer-plugins';

const atomizer = rollup({
    /* options */
    config: atomizerConfig,
    outfile: 'dist/atomizer.css',
});

export default {
    // ... rollup config
    plugins: [atomizer],
};
```

Configuration information and examples are available in the [`atomizer-plugins` repository](https://github.com/acss-io/atomizer/tree/main/packages/atomizer-plugins).

## Start your build process

Run your build setup as configured in your project's `package.json`.

```shell
npm run dev
```

## Begin using Atomizer

Add the generated Atomizer CSS to your template, then start adding Atomizer classes to your markup.

```html
<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <link rel="stylesheet" href="/dist/atomizer.css" />
    </head>
    <body>
        <h1 class="Fw(b) Fz(2rem)">Welcome!</h1>
    </body>
</html>
```
