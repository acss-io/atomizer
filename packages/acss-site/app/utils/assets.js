'use strict';

var objectAssign = require('object-assign');

var assets = require('../build/assets.json');
var images = require('../build/images.json');
var manifest = objectAssign(assets, images);

module.exports = manifest;
