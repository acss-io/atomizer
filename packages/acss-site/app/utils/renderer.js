var marked = require('marked');
var renderer = new marked.Renderer();

renderer.heading = function (text, level) {
    var escapedText = text.toLowerCase().replace(/[^\w]+/g, '-');

    return (
        '<h' + level + '>' +
        '<a name="' + escapedText + '" class="anchor"></a>' +
        text + ' ' +
        '<a href="#' + escapedText + '" class="hash-link">#</a>' +
        '</h' + level + '>'
    );
};

module.exports = renderer;
