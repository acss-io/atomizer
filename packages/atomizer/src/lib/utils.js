const _ = require('lodash');

const utils = {};
const customValueTokenRegex = /#\{(.+?)\}/g;

// hex value to rgb object
utils.hexToRgb = function (hex /*:string*/) /*:Rgb*/ {
    // shorthand to full form
    hex = hex.replace(/^#?([a-f\d])([a-f\d])([a-f\d])$/i, function (m, r, g, b) {
        return r + r + g + g + b + b;
    });

    const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
    return result
        ? {
              r: parseInt(result[1], 16),
              g: parseInt(result[2], 16),
              b: parseInt(result[3], 16),
          }
        : null;
};

/**
 * helper function to handle merging array of strings
 * @param  {mixed} a Data of the first merge param
 * @param  {mixed} b Data of the second merge param
 * @return {mixed}   The merged array
 */
utils.handleMergeArrays = function (a, b) {
    if (_.isArray(a) && _.isArray(b)) {
        return _.union(a, b).sort();
    }
};

// merge atomizer configs into a single config
utils.mergeConfigs = function (configs /*:Config[]*/) /*:Config*/ {
    // TODO: Offer option to warn on conflicts
    return _.mergeWith.apply(null, configs.concat(utils.handleMergeArrays));
};

// returns a repeated string by X amount
utils.repeatString = function (pattern /*:string*/, count /*:integer*/) {
    let result = '';
    if (count < 1) {
        return result;
    }
    while (count > 1) {
        if (count & 1) {
            result += pattern;
        }
        (count >>= 1), (pattern += pattern);
    }
    return result + pattern;
};

// Performs a lookup of custom value tokens, recursively replacing
//  any custom value tokens found within each custom value
utils.getCustomValue = function (config /*:Config*/, currentName /*:string*/, nameStack /*:string[]*/) /*:string*/ {
    nameStack = nameStack || [];

    if (typeof currentName !== 'string' || !config || !config.custom) {
        return null;
    }

    // If not found, return null, which will prevent the rule
    // from being written to CSS
    const value = config.custom[currentName] || null;

    // Short circuit if value isn't a string, or doesn't contain
    // any tokens that need to be replaced
    if (typeof value !== 'string' || value.indexOf('#{') === -1) {
        return value;
    }

    // Add this custom value name to our breadcrumb trail
    nameStack.push(currentName);

    // Limit the depth of recursion to help avoid infinite loops
    // Expectation is that 20 should be a more than reasonble depth
    // to assume something is wrong
    if (nameStack.length > 20) {
        throw new Error(
            `Depth limit reached while substituting custom value tokens. Ensure your custom values don't contain tokens that reference one another, leading to an infinite loop.\n\nCustom value trace: ${nameStack.join(
                ' > '
            )}`
        );
    }

    return value.replace(customValueTokenRegex, (token, name) => {
        return this.getCustomValue(config, name, nameStack);
    });
};

module.exports = utils;
