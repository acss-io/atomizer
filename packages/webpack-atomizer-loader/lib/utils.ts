import * as fs from 'fs';
import * as path from 'path';

const PATH_SEP: string = '/';

export function writeCssFile(cssDest: string, cssString: string): Promise<Function> {
    return new Promise((resolve, reject) => {
        fs.writeFile(cssDest, cssString, err => {
            if (err) {
                reject(err);
            } else {
                resolve();
            }
        });
    });
}

export function ensureFolderExists(path: string, mask: string = '0777'): boolean {
    let isExisted: boolean = false;
    try {
        fs.mkdirSync(path, mask);
        isExisted = true;
    } catch (err) {
        if (err.code === 'EEXIST') {
            isExisted = true;
        }
    }
    return isExisted;
}

export function ensureExists(filePath: string): boolean {
    let dirs: string[] = path.dirname(filePath).split(PATH_SEP);
    let result: boolean = true;
    let currentPath: string;

    if (dirs[0] === '') {
        dirs[0] = path.sep;
    }

    dirs.forEach((_, i, p) => {
        currentPath = path.join.apply(null, p.slice(0, i + 1));
        if (!ensureFolderExists(currentPath)) {
            result = false;
        }
    });
    return result;
}
