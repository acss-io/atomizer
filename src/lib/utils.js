var _ = require('lodash');

var utils = {};

// hex value to rgb object
utils.hexToRgb = function (hex/*:string*/)/*:Rgb*/ {
    var result;

    // shorthand to full form
    hex = hex.replace(/^#?([a-f\d])([a-f\d])([a-f\d])$/i, function(m, r, g, b) {
        return r + r + g + g + b + b;
    });

    result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
    return result ? {
        r: parseInt(result[1], 16),
        g: parseInt(result[2], 16),
        b: parseInt(result[3], 16)
    } : null;
};

/**
 * helper function to handle merging array of strings
 * @param  {mixed} a Data of the first merge param
 * @param  {mixed} b Data of the second merge param
 * @return {mixed}   The merged array
 */
utils.handleMergeArrays = function (a, b) {
    if (_.isArray(a) && _.isArray(b)) {
        return _.union(a, b).sort();
    }
};

// merge atomizer configs into a single config
utils.mergeConfigs = function (configs/*:Config[]*/)/*:Config*/ {
    // TODO: Offer option to warn on conflicts
    return _.mergeWith.apply(null, configs.concat(utils.handleMergeArrays));
};

// returns a repeated string by X amount
utils.repeatString = function (pattern/*:string*/, count/*:integer*/) {
    var result = '';
    if (count < 1) {
        return result;
    }
    while (count > 1) {
        if (count & 1) {
            result += pattern;
        }
        count >>= 1, pattern += pattern;
    }
    return result + pattern;
};

module.exports = utils;
