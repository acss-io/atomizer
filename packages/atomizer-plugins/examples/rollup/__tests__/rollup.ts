import { rollup } from 'rollup';
import { readFileSync } from 'fs';
import { join } from 'path';
import rollupConfig from '../rollup.config.mjs';
import { getDirname } from '../../../../../tests/utils.mjs';

const __dirname = getDirname(import.meta.url);
const atomizerCssPath = join(__dirname, '..', 'dist', 'atomizer.css');

describe('rollup', () => {
    it('should create atomizer css', async () => {
        const bundle = await rollup(rollupConfig);
        await bundle.write({
            file: join(__dirname, '..', 'dist', 'main.js'),
            format: 'esm',
        });
        await bundle.close();

        // assert that the css file was created
        const css = readFileSync(atomizerCssPath, 'utf8');
        expect(css).toMatchSnapshot();
    });
});
