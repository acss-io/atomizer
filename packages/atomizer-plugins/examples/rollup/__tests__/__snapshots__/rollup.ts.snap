// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rollup should create atomizer css 1`] = `
".Bdrs\\\\(\\\\.5rem\\\\) {
  border-radius: .5rem;
}
.Bgc\\\\(backgroundColor\\\\) {
  background-color: #333;
}
.C\\\\(textColor\\\\) {
  color: #fff;
}
.Ff\\\\(ss\\\\) {
  font-family: Helvetica, Arial, sans-serif;
}
.Fw\\\\(b\\\\) {
  font-weight: bold;
}
.Fz\\\\(2rem\\\\) {
  font-size: 2rem;
}
.Lh\\\\(1\\\\.5\\\\) {
  line-height: 1.5;
}
.Mx\\\\(a\\\\) {
  margin-left: auto;
  margin-right: auto;
}
.Px\\\\(1rem\\\\) {
  padding-left: 1rem;
  padding-right: 1rem;
}
.Py\\\\(\\\\.1rem\\\\) {
  padding-top: .1rem;
  padding-bottom: .1rem;
}
.W\\\\(35vw\\\\) {
  width: 35vw;
}
"
`;
